README.md
# VSCode - Enum Generator

An extension to generate enums with variable number of members and values with a defined hamming distance between them

![animatedScreenshot](https://github.com/abdalmoniem/VSCode-Enum-Generator/raw/main/media/aimatedScreenshot.gif)

## Features
create custom enums with defined bit width, member count and, hamming distance by typing in `e[bw]c[mc]h[hd]` where:
   
   1. bw: bit width
   2. mc: member count
   3. hd: hamming distance


so `e32c10h3` would generate a 32-bit enum with 10 members and a hamming distance of 3 between them:

```c
typedef enum {
   MEMBER_1 = 0xdda458cf,
   /* hamming distance is 17 */
   MEMBER_2 = 0x3e8ccab,
   /* hamming distance is 19 */
   MEMBER_3 = 0x8fc3fa2a,
   /* hamming distance is 14 */
   MEMBER_4 = 0x5d1750d6,
   /* hamming distance is 17 */
   MEMBER_5 = 0x2653c801,
   /* hamming distance is 13 */
   MEMBER_6 = 0x50fac185,
   /* hamming distance is 11 */
   MEMBER_7 = 0xa400be4d,
   /* hamming distance is 18 */
   MEMBER_8 = 0xfc637176,
   /* hamming distance is 15 */
   MEMBER_9 = 0x62e1be02,
   /* hamming distance is 16 */
   MEMBER_10 = 0x3abf37f3
} ENUM_NAME;
``` 

note that the hamming number generator generates random numbers and ensures minimum specified hamming distance between them (so in case of the previous example, it would generate numbers with minimum distance of 3 and maximum distance of `bitWidth`-1), so you're not likely getting the same enum values twice anytime soon.

CHANGELOG.md
# Change Log

All notable changes to the "enum-hamming-distance-generator" extension will be documented in this file.

## [1.3.0]

- Initial release

src/hammingNumberGenerator.ts
/**
 * file: hammingNumberGenerator.js
 * 
 * date: 08-Apr-22
 * 
 * author: AbdAlMoniem AlHifnawy
 * 
 * description: Generates an array of numbers of configurable length and configurable
 *                hamming distance between its items
 */

export enum Integers {
   BINARY = 2,
   DECIMAL = 10,
   HEXADECIMAL = 16
}

export async function generateCodebook(
   bitWidth: number,
   size: number,
   hammingDistance: number,
   representation: Integers,
   loopIterations: number,
   allowReduction: boolean,
   autoGenerate: boolean,
   timeout: number): Promise<{ hammingCodes: Array<string> | undefined, minHam: number | undefined }> {
   if (hammingDistance >= bitWidth) {
      return { hammingCodes: undefined, minHam: undefined };
   }

   const seen: Array<string> = new Array();
   const hammingCodes: Array<string> = new Array();
   let iterations = loopIterations;
   let currentHammingDistance = hammingDistance;

   let start = Date.now();

   while (true) {
      /*********************** exit conditions ********************/
      if ((hammingCodes.length >= size) && !autoGenerate) { break; }
      if (timeout > 0) {
         if (Date.now() - start >= timeout) {
            if (!allowReduction) { break; }
            currentHammingDistance -= 1;

            start = Date.now();
         }
      } else {
         iterations -= 1;
         if (iterations <= 0) {
            if (!allowReduction) { break; }
            currentHammingDistance -= 1;
            iterations = loopIterations;
         }
      }
      if (currentHammingDistance <= 0) { break; }

      let testCandidate = getRandomBinStr(bitWidth);
      if (seen.includes(testCandidate)) { continue; }
      seen.push(testCandidate);

      if (hammingCodes.length === 0) { 
         hammingCodes.push(testCandidate);
      } else {
         let valid = true;
         for (let word of hammingCodes) {
            if ((testCandidate === word) || await calculateHammingDistance(testCandidate, word) < currentHammingDistance) {
               valid = false;
               break;
            }
         }

         if (valid) { hammingCodes.push(testCandidate); }
      }
   }

   if (hammingCodes.length > 0) {
      const minimums: Array<number> = new Array();

      for (let index = 0; index < hammingCodes.length - 1; index++) {
         minimums.push(await calculateHammingDistance(hammingCodes[index], hammingCodes[index + 1]));
      }

      let minHam = minimums[0];

      for (let item of minimums) {
         if (minHam < item) { minHam = item; }
      }

      if (hammingCodes.length > 0) {
         minHam = currentHammingDistance;
      }

      if (hammingCodes.length !== size) {
         console.error(`could not find more than ${hammingCodes.length} code(s) with minimum hamming distance of ${minHam} between them`);
      }

      if (currentHammingDistance !== hammingDistance) {
         console.error(`could not generate ${hammingCodes.length} code(s) with minimum hamming distance of more than ${minHam} between them`);
      }

      for (let code of hammingCodes.splice(0, hammingCodes.length)) {
         hammingCodes.push(convertBinarCodeToNumberString(code, bitWidth, representation) as string);
      }

      return { hammingCodes, minHam };
   } else {
      console.error('Could not generate any codes with the given parameters');
   }

   return { hammingCodes: undefined, minHam: undefined };
}
export function getRandomBinStr(width: number): string {
   let binaryString = '';
   for (let i = 0; i < width; i++) {
      const randomBit = Math.floor(Math.random() * 2);

      binaryString += randomBit.toString();
   }

   return binaryString;
}

export async function calculateHammingDistance(s1: string, s2: string): Promise<number> {
   const n1 = parseBigInt(s1);
   const n2 = parseBigInt(s2);

   let onesCount = 0;

   for (let bit of await binary((n1 as bigint) ^ (n2 as bigint))) {
      if (bit === '1') { onesCount += 1; }
   }

   return onesCount;
}

export function convertBinarCodeToNumberString(code: string, bitWidth: number, representation: Integers): string | undefined {
   if (representation === Integers.BINARY) { return code.padStart(bitWidth, '0'); }
   else if (representation === Integers.DECIMAL) { return `${parseBigInt(code)}`; }
   else if (representation === Integers.HEXADECIMAL) { return (`0x${parseBigInt(code).toString(16)}`).padStart(bitWidth / 4, '0'); }
}

export async function binary(dec: BigInt): Promise<string> {
   if (dec >= BigInt(0)) {
      return dec.toString(2);
   } else {
      /* Here you could represent the number in 2s compliment but this is not what 
         JS uses as its not sure how many bits are in your number range. There are 
         some suggestions https://stackoverflow.com/questions/10936600/javascript-decimal-to-binary-64-bit 
      */
      return (~dec).toString(2);
   }
}

export function parseBigInt(binStr: string): BigInt {
   const lastIndex = binStr.length - 1;
   let total = BigInt(0);

   for (let i = 0; i < binStr.length; i++) {
      if (binStr[lastIndex - i] === '1') {
         total += (BigInt(2) ** BigInt(i));
      }
   }

   return total;
}

src/extension.ts
/**
 * file: extension.js
 * 
 * date: 08-Apr-22
 * 
 * author: AbdAlMoniem AlHifnawy
 * 
 * description: main extension file that executes extension functionalities
 */

import * as vscode from 'vscode';
import { Integers, generateCodebook, calculateHammingDistance } from './hammingNumberGenerator';

const enumRegex = /typedef\s*enum\s*{\s*(?:(?<memberName>\w+)\s*=?\s*(?<memberValue>(?:0b[01]+)|(?:0x[a-fA-F0-9]+)|(?:\d+))?,?\s*)+\s*}\s*\w+;/m;

export function activate(context: vscode.ExtensionContext) {
	const enumCompletionItemProvider = new EnumCompletionItemProvider();

	context.subscriptions.push(
		vscode.languages.registerCompletionItemProvider('c', enumCompletionItemProvider),
		vscode.languages.registerCompletionItemProvider('cpp', enumCompletionItemProvider),

		vscode.commands.registerCommand('enum-generator.convertEnum',
			async () => await convertEnum()),
		vscode.commands.registerCommand('enum-generator.generateEnum',
			async (bitWidth, memberCount, hammingDistance) => await generateEnum(bitWidth, memberCount, hammingDistance)),

		vscode.window.onDidChangeTextEditorSelection(event =>
			vscode.commands.executeCommand('setContext',
				'canConvert',
				enumRegex.test(
					event.textEditor.document.getText(
						event.textEditor.selection))))
	);
}

// this method is called when your extension is deactivated
export function deactivate() { }

export async function convertEnum() {
	const editor = vscode.window.activeTextEditor;
	if (editor) {
		const document = editor.document;
		const selection = editor.selection;
		const selectionText = document.getText(selection);

		if (enumRegex.test(selectionText)) {
			vscode.window.showInformationMessage('Valid Enum, Converting...');
			let selectionTextLines = selectionText.split(/[\r\n]/);

			if (selectionTextLines.length === 1) {
				const lineRegex = /(?<enumHeader>typedef\s*enum\s*{)(?<members>\s*(?:(?<memberName>\w+)\s*=?\s*(?<memberValue>(?:0b[01]+)|(?:0x[a-fA-F0-9]+)|(?:\d+))?,?\s*)+\s*)(?<enumFooter>}\s*\w+;)/;

				const match = lineRegex.exec(selectionTextLines[0]);
				const members = match!.groups?.members.split(',').filter(value => value.trim().length > 0);

				let convertedEnum = match!.groups?.enumHeader;

				const { hammingCodes, minHam } = await generateCodebook(16, members!.length, 3,
					Integers.HEXADECIMAL,
					3000,
					false,
					false,
					300);

				if (hammingCodes !== undefined) {
					members!.forEach((member, index) => {
						let groups = /(?<filler1>.*?)(?<memberName>\w+)(?<filler2>\s*=?\s*)(?<memberValue>(?:0b[01]+)|(?:0x[a-fA-F0-9]+)|(?:\d+))?(?<filler3>.*)/.exec(member)?.groups;

						convertedEnum += `${groups?.filler1}${groups?.memberName}${groups!.filler2.trim().length > 0 ? groups?.filler2 : '='}${hammingCodes[index]}${groups?.filler3}${index < members!.length - 1 ? ',' : ''}`;
					});

					convertedEnum! += match!.groups?.enumFooter;

					editor.edit(editBuilder => editBuilder.replace(selection, convertedEnum!));
				}
			}

			let memberCount = 0;
			for (let line of selectionTextLines) {
				if (/^(\s*)(?!typedef)(?<memberName>\w+)\s*=?\s*(?<memberValue>(?:0b[01]+)|(?:0x[a-fA-F0-9]+)|(?:\d+))?/.test(line)) {
					memberCount += 1;
				}
			}

			const { hammingCodes, minHam } = await generateCodebook(16, memberCount, 3,
				Integers.HEXADECIMAL,
				3000,
				false,
				false,
				300);

			if (hammingCodes !== undefined) {
				let convertedEnum = '';
				let codeIndex = 0;
				selectionTextLines.forEach((line, index) => {
					const lineRegex = /^(\s*)(?!typedef)(?<memberName>\w+)\s*=?\s*(?<memberValue>(?:0b[01]+)|(?:0x[a-fA-F0-9]+)|(?:\d+))?/;
					if (lineRegex.test(line)) {
						convertedEnum += `${line.replace(lineRegex, `$1$2 = ${hammingCodes[codeIndex]}`)}\n`;
						codeIndex += 1;
					} else {
						convertedEnum += `${line}${index < selectionTextLines.length - 1 ? '\n' : ''}`;
					}
				});
				editor.edit(editBuilder => editBuilder.replace(selection, convertedEnum));
			}
		} else {
			vscode.window.showInformationMessage('Invalid Enum');
		}
	}
}

export async function generateEnum(bitWidth: number, memberCount: number, hammingDistance: number): Promise<void> {
	if (bitWidth === undefined) {
		let answer = await vscode.window.showInputBox(
			{
				prompt: 'Bit Width',
				title: 'Enter Bit Width',
				placeHolder: 'Enter Bit Width (ex. 128, 64, 32, etc...)',
				ignoreFocusOut: true,
				validateInput: text => /^\d+$/.test(text) ? null : 'not a valid number!'
			});

		if (answer) {
			bitWidth = parseInt(answer);
		}
	}

	if (memberCount === undefined) {
		let answer = await vscode.window.showInputBox(
			{
				prompt: 'Member Count',
				title: 'Enter Number of Members',
				ignoreFocusOut: true,
				placeHolder: 'Enter Number of Members (ex. 4, 5, 6, etc...)',
				validateInput: text => /^\d+$/.test(text) ? null : 'not a valid number!'
			});

		if (answer) {
			memberCount = parseInt(answer);
		}
	}

	if (hammingDistance === undefined) {
		let answer = await vscode.window.showInputBox(
			{
				prompt: 'Hamming Distance',
				title: 'Enter Hamming Distance',
				ignoreFocusOut: true,
				placeHolder: 'Enter Hamming Distance (ex. 2, 3, 4, etc...)',
				validateInput: text => /^\d+$/.test(text) ? null : 'not a valid number!'
			});

		if (answer) {
			hammingDistance = parseInt(answer);
		}
	}

	const editor = vscode.window.activeTextEditor;

	if (editor) {
		const { hammingCodes, minHam } = await generateCodebook(bitWidth, memberCount, hammingDistance,
			Integers.HEXADECIMAL,
			3000,
			false,
			false,
			300);

		// console.log({ hammingCodes, minHam });
		if (hammingCodes !== undefined) {
			let snippet = '';
			for (let i = 0; i < memberCount; i++) {
				snippet += `\t\${${i + 1}:MEMBER_${i + 1}} = ${hammingCodes[i] !== undefined ? hammingCodes[i] : 0x00}${i < memberCount - 1 ? ',' : ''}\n`;
				if ((hammingCodes[i] !== undefined) && (hammingCodes[i + 1] !== undefined)) {
					if (i < memberCount - 1) {
						snippet += `\t/* hamming distance is ${await calculateHammingDistance(
							BigInt(hammingCodes[i]).toString(2),
							BigInt(hammingCodes[i + 1]).toString(2))} */\n`;
					}
				}
			}
			snippet += `} \${${memberCount + 1}:ENUM_NAME};`;

			await editor.insertSnippet(new vscode.SnippetString(snippet));
		}
	}
}

export class EnumCompletionItemProvider implements vscode.CompletionItemProvider {
	async provideCompletionItems(
		document: vscode.TextDocument,
		position: vscode.Position,
		token: vscode.CancellationToken,
		context: vscode.CompletionContext): Promise<vscode.CompletionList<vscode.CompletionItem> | vscode.CompletionItem[]> {
		const completionItems: vscode.CompletionList<vscode.CompletionItem> = new vscode.CompletionList();
		const linePrefix = document.lineAt(position).text.substring(0, position.character);

		const regex = /e(?<bitWidth>\d+)(?:c(?<memberCount>\d+)?(?:h(?<hammingDistance>\d+)?)?)?/;

		if (regex.test(linePrefix)) {
			const match = regex.exec(linePrefix);
			const bitWidth = parseInt(match?.groups?.bitWidth!);
			const memberCount = parseInt(match?.groups?.memberCount!);
			const hammingDistance = parseInt(match?.groups?.hammingDistance!);

			if (isNaN(memberCount)) {
				for (let mc = 2; mc < 5; mc++) {
					for (let hd = 2; hd < (10 > bitWidth ? bitWidth - 1 : 10); hd++) {
						completionItems.items.push.apply(completionItems.items, await this.getSnippets(bitWidth, mc, hd, match!.input));
					}
				}
			} else if (isNaN(hammingDistance)) {
				for (let hd = 2; hd < (10 > bitWidth ? bitWidth - 1 : 10); hd++) {
					completionItems.items.push.apply(completionItems.items, await this.getSnippets(bitWidth, memberCount, hd, match!.input));
				}
			} else {
				completionItems.items.push.apply(completionItems.items, await this.getSnippets(bitWidth, memberCount, hammingDistance, match!.input));
			}

			completionItems.isIncomplete = true;
		}

		// return all completion items as an array
		return completionItems;
	}

	private async getSnippets(bitWidth: number, memberCount: number, hammingDistance: number, filterText: string): Promise<Array<vscode.CompletionItem>> {
		const completionItems: Array<vscode.CompletionItem> = new Array();

		const enumSnippet = new vscode.CompletionItem(`${bitWidth}-bit enum count ${memberCount}, dis ${hammingDistance}`);
		enumSnippet.detail = `${memberCount} member ${bitWidth}-bit Enum`;
		enumSnippet.insertText = 'typedef enum {\n';
		enumSnippet.filterText = filterText;
		enumSnippet.kind = vscode.CompletionItemKind.Snippet;
		enumSnippet.documentation =
			new vscode.MarkdownString(`Inserts a ${bitWidth}-bit enum snippet with
						${memberCount} members and minimum hamming distance of ${hammingDistance} between them.`);
		enumSnippet.command = { command: 'enum-generator.generateEnum', title: 'hello', arguments: [bitWidth, memberCount, hammingDistance, enumSnippet] };

		completionItems.push(enumSnippet);

		return completionItems;
	}
}